### PR TITLE
Add functionaly for SMIRNOFF convention of permuting atoms for improper calculations

### LINKED ISSUE #101: Add SMIRNOFF improper handling
Currently, espaloma implements a six-fold improper torsion scheme (see image) that includes pairs of improper torsions that mostly, but not completely, cancel out:
image
This differs from the SMIRNOFF improper torsion definition, which includes three impropers in a consistent ordering to ensure they gracefully handle non-planar restraints.

We should implement an additional mode with the SMIRNOFF improper definition, and see how things change when we refit the espaloma model.

This will also allow us to export the model to OpenMM via `openmmforcefields`:

cc: https://github.com/openmm/openmmforcefields/pull/182#issuecomment-998448457

### PR DESCRIPTION
**Pull request by kaminow**
Added functionality to use SMIRNOFF convention for impropers instead of the current espaloma convention. Current default options will use the SMIRNOFF convention for newly created molecules. When datasets are loaded, no changes will be made if using the default options, however if an argument for `improper_def` is provided, the `n4_improper` nodes will be regenerated according to the provided input.

Resolves #101.

### CODE REVIEW

File: espaloma/graphs/graph.py [hunk 1]
```diff
@@ -44,7 +44,8 @@ class Graph(BaseGraph):
 
     """
 
-    def __init__(self, mol=None, homograph=None, heterograph=None):
+    def __init__(self, mol=None, homograph=None, heterograph=None,
```
Comment (yuanqing-wang) [inline espaloma/graphs/graph.py:47]: this modification is a bit too aggressive as we need to supply another arguments at the very beginning stage of graph definition. can we not touch the `esp.graph` part and move this to another utility function where we take a `esp.Graph` and modify in-place?

File: espaloma/data/dataset.py [hunk 2]
```diff
@@ -340,14 +340,14 @@ def save(self, path):
             graph.save(path + "/" + str(idx))
 
     @classmethod
-    def load(cls, path):
+    def load(cls, path, improper_def=None):
```
Comment (yuanqing-wang) [inline espaloma/data/dataset.py:343]: same as in `esp.graph`. looks like this is changing too much of the mainline code

File: espaloma/nn/readout/janossy.py [hunk 3]
```diff
@@ -197,6 +198,9 @@ def __init__(
         # bookkeeping
         self.out_features = out_features
         self.levels = ["n4_improper"]
+        if improper_def not in ('smirnoff', 'espaloma'):
```
Comment (yuanqing-wang) [inline espaloma/nn/readout/janossy.py:201]: might be nice moving this to another object called `JanossyPoolingWithSmirnoffImproper` or whatever

**Review by yuanqing-wang — CHANGES_REQUESTED** [review]
thanks for this PR! I think this would do exactly what we wanted to do!

only comment is that it would be nice if we can move the heavy lifting to a utility script and keep the `esp.Graph` unchanged

**Comment by jchodera** [comment]
> only comment is that it would be nice if we can move the heavy lifting to a utility script and keep the esp.Graph unchanged

Which "heavy lifting" are you referring to specifically, @yuanqing-wang ? Do you just mean the Janossy pooling variant schemes should be moved to objects/classes?

**Comment by yuanqing-wang** [comment]
> > only comment is that it would be nice if we can move the heavy lifting to a utility script and keep the esp.Graph unchanged
> 
> Which "heavy lifting" are you referring to specifically, @yuanqing-wang ? Do you just mean the Janossy pooling variant schemes should be moved to objects/classes?

at this stage, I think the enumeration of the improper indices should be moved to a utility function as opposed to handled in the `esp.Graph` object

**Comment by yuanqing-wang** [comment]
It would be great to move it to graph.utils

**Review by yuanqing-wang — APPROVED** [review]

**Pull request merged**